pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+used in existing typed and untyped open-source code. We determined
+that the vast majority of use cases are covered.
+
+We considered adding support for named, optional, and variadic
+arguments. However, we decided against including these features, as
+our analysis showed they are infrequently used. When they are really
+needed, it is possible to type these using `Callback Protocols
+<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.
+
+See the Rejected Alternatives section for more detailed discussion
+about omitted features.
+
+Specification
+=============
+
+Typing Behavior
+---------------
+
+Type checkers should treat the new syntax with exactly the same
+semantics as ``typing.Callable``.
+
+As such, a type checker should treat the following pairs exactly the
+same::
+
+    from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple
+
+    P = ParamSpec("P")
+    Ts = TypeVarTuple('Ts')
+
+    f0: () -> bool
+    f0: Callable[[], bool]
+
+    f1: (int, str) -> bool
+    f1: Callable[[int, str], bool]
+
+    f2: (...) -> bool
+    f2: Callable[..., bool]
+
+    f3: async (str) -> str
+    f3: Callable[[str], Awaitable[str]]
+
+    f4: (**P) -> bool
+    f4: Callable[P, bool]
+
+    f5: (int, **P) -> bool
+    f5: Callable[Concatenate[int, P], bool]
+
+    f6: (*Ts) -> bool
+    f6: Callable[[*Ts], bool]
+
+    f7: (int, *Ts, str) -> bool
+    f7: Callable[[int, *Ts, str], bool]
+
+
+Grammar and AST
+---------------
+
+The proposed new syntax can be described by these AST changes ::
+
+    expr = <prexisting_expr_kinds>
+         | AsyncCallableType(callable_type_arguments args, expr returns)
+         | CallableType(callable_type_arguments args, expr returns)
+
+    callable_type_arguments = AnyArguments
+                            | ArgumentsList(expr* posonlyargs)
+                            | Concatenation(expr* posonlyargs, expr param_spec)
+
+
+Here are our proposed changes to the `Python Grammar
+<https://docs.python.org/3/reference/grammar.htm>`::
+
+    expression:
+        | disjunction disjunction 'else' expression
+        | callable_type_expression
+        | disjunction
+        | lambdef
+
+    callable_type_expression:
+        | callable_type_arguments '->' expression
+        | ASYNC callable_type_arguments '->' expression
+
+    callable_type_arguments:
+        | '(' '...' [','] ')'
+        | '(' callable_type_positional_argument*  ')'
+        | '(' callable_type_positional_argument* callable_type_param_spec ')'
+
+    callable_type_positional_argument:
+        | !'...' expression ','
+        | !'...' expression &')'
+
+    callable_type_param_spec:
+        | '**' expression ','
+        | '**' expression &')'
+
+
+
+If PEP 646 is accepted, we intend to include support for unpacked
+types by modifying the grammar for
+``callable_type_positional_argument`` as follows::
+
+    callable_type_positional_argument:
+        | expression ','
+        | expression &')'
```

Comment:
What happened to the `!'...'` guard?

Reply:
Whoops - this section predates me realizing we needed that guard. Will fix!